leapp/libraries/common/call.py

```diff
@@ -0,0 +1,135 @@
+
+
+STDOUT = sys.stdout.fileno()
+STDERR = sys.stderr.fileno()
+
+
+def decode(data):
+    return data.decode('utf-8')
+
+
+def multiplex_reads(ep, fds, callback_raw, callback_linebuffered, timeout=1, buffer_size=72):
+    # Register the file descriptors (stdout + stderr) with the epoll object
+    # so that we'll get notifications when data are ready to read
+    for fd in fds:
+        ep.register(fd, select.EPOLLIN | select.EPOLLPRI)
+
+    # We need to keep track of which file descriptors have already been "drained"
+    hupped = set()
+    # Set up file-descriptor specific buffers where we'll buffer the output
+    buf = {fd: bytes() for fd in fds}
+    linebufs = {fd: '' for fd in fds}
+
+    while not ep.closed and len(hupped) != len(fds):
+        events = ep.poll(timeout)
+        for fd, event in events:
+            if event == select.EPOLLHUP:
+                hupped.add(fd)
+            read = os.read(fd, buffer_size)
+            callback_raw(fd, read)
+            linebufs[fd] += decode(read)
+            if '\n' in linebufs[fd]:
+                pre, post = linebufs[fd].split('\n', 1)
+                linebufs[fd] = post
+                callback_linebuffered(fd, pre)
+            buf[fd] += read
+
+    # Process leftoevers from line buffering
+    for (fd, lb) in linebufs.items():
+        if lb:
+            callback_linebuffered(fd, lb)
+
+    return buf
+
+
+def _call(command, callback_raw=lambda fd, value: None, callback_linebuffered=lambda fd, value: None,
+          encoding='utf-8', poll_timeout=1, read_buffer_size=42):
+    ''' Call a command and return it's output
+
+    Args:
+        command          - the command to execute
+        encoding         - decode output using this encoding
+        poll_timeout     - timeout used by epoll to wait certain amount of time for activity on
+                           file descriptors
+        read_buffer_size - how much data are we going to read form the file descriptors each
+                           iteration
+
+        callback_raw          - callback executed on data as they are read from file descriptors
+        callback_linebuffered - callback executed on lines as they are read from the file descriptors
+    '''
+    if not isinstance(command, list):
+        raise TypeError('command parameter has to be a list')
+
+    # Create a separate pipe for stdout/stderr
+    #
+    # The parent process is going to use the read-end of the pipes for reading child's
+    # stdout/stderr, whereas the forked children process is going to use the write-end
+    # of the pipes to pass data to parent
+    stdout, wstdout = os.pipe()
+    stderr, wstderr = os.pipe()
+    ep = select.epoll()
+
+    pid = os.fork()
+    if pid > 0:
+        # Since pid > 0, we are in the parent process, so we have to close the write-end
+        # file descriptors
+        os.close(wstdout)
+        os.close(wstderr)
+
+        read = multiplex_reads(
+            ep,
+            [stdout, stderr],
```

Comment:
I guess that it won't matter that much, but you never change this list, I think it's a good idea to have it a tuple, lists are just noice IMHO but well I will let this be up to you.